### DISCUSSION TITLE
v1.0.0 Beta

### DISCUSSION THREAD

**Discussion by ghost** [BerriAI/litellm#648] [Ideas]
Starting a discussion to track potential refactoring / breaking changes we need to make for a v1. 

**Motivation** 
Bundle near-term breaking changes into 1 release.

Driven by the discussion on https://github.com/BerriAI/litellm/issues/636. 

**Additional Thoughts**
I haven't had a chance to explore the new openai sdk, but I anticipate that's going to cause some issues too. https://github.com/openai/openai-python/discussions/631

**Timeline**
We should aim to have a locked in list by 30th October.

**Comment by ghost** [comment]
https://github.com/BerriAI/litellm/issues/636 - 
"I would suggest to make get_max_tokens return only an integer representig the max_tokens of a given model, and maybe add a get_model_info function returning the currenct dict with all the info!" - @mistrello96

**Reply by mistrello96** [reply]
Hi, has this been addressed in the 1.0.0 release? It seems the function still returns a dict

**Reply by ishaan-jaff** [reply]
yes @mistrello96 - addressed here: https://github.com/BerriAI/litellm/commit/bd825595539f38fe1d8b97ad6e0e172baeec7066

**Reply by ishaan-jaff** [reply]
migration guide: https://docs.litellm.ai/docs/migration

**Comment by ghost** [comment]
remove old debugger dashboard variables - https://github.com/BerriAI/litellm/blob/cb100d19dadcdbb6494dd961125c5600ec892bcb/litellm/__init__.py#L10

**Comment by ghost** [comment]


**Reply by octopop** [reply]
What will supersede the CLI proxy?

**Reply by lzumot** [reply]
would be nice to retain that for those of us who are not using it for production but still benefit from the cli proxy.

**Reply by vividfog** [reply]
The CLI proxy is the killer app part of LiteLLM for me. I hope there's some plan B for it. It does not need to be a proxy server for production. What there is works perfectly well for local installations. Universal OpenAI-style API to all local models is the use case I've been promoting LiteLLM for, and I don't know what to replace it with, if this feature is gone. For instance, Ollama doesn't have an OpenAI API, and LiteLLM is excellent for making Ollama look like OpenAI.

**Reply by ghost** [reply]
Hey the docker image should still solve the issue - https://docs.litellm.ai/docs/simple_proxy#usage.

**Reply by ghost** [reply]
Hey @octopop @lzumot @vividfog seems like we missed a big use-case of the cli which was eval. I want to make sure we don't release a 'worse' experience. 

I would love to understand this better from your side: https://calendly.com/d/4mp-gd3-k5k/litellm-1-1-onboarding-chat

**Reply by ishaan-jaff** [reply]
Thanks to everyone for sharing their thoughts on the proxy - we'd love to learn more about how we can make it better for your use case, sharing my cal here if someone has a chance to share how they're using the proxy: https://calendly.com/ishaan-berri/30min?month=2023-11 

if call does not work comment in this thread how you're using it

**Reply by shuther** [reply]
Proxy/router is the killer feature for me even if I am still confused how to use it (litellm_server vs litellm/proxy vs litellm/__init). Maybe I need a good example with poetry and where to save the secrets/configuration so we can remove legacy information in the documentation.
It is important so we can consider there is ONE api standard to call an LLM (openai) and we can leverage workflows built around openAI for best-of-the-breed models through litellm. Should it be a transparent proxy in some situation or acting as a API gateway and balancing to different providers/models based on specific strategy is exactly where I see a spot for it. my 2 cents.

**Reply by ghost** [reply]
@shuther let me know what was missing in the docs - https://docs.litellm.ai/docs/simple_proxy

Would appreciate a quick chat to understand your scenario better - https://calendly.com/d/4mp-gd3-k5k/litellm-1-1-onboarding-chat

**Reply by shuther** [reply]
I am based in europe, in which timezone are you so we can orgwnize this call if it is easier?
Few of my questions:
- there are multiple DockerFile/.env files within the project; what are the differences? As a general rule I am confused because there are "sub" packages called proxy and litlellm_server but the documentation states we should run: litellm for everything  (so we should never call proxy_cli directly). is it correct? I refer here to a dev environment.
- there are some references to a file called litellm.secrets.toml but I can't find an example.
- I would like litlellm to behave as such: - is it how it is expected to work (should I refer to the proxy or router documentation)?
  1.  Ignore the API key from the request, keep the model or work with a fallback one (it should retry in case we face rate limitations...);
  2.  identify an API key based on a configuration file/strategy 
  3. forward the request to OpenAI (i.e.) 
  4. retry if needed

- the examples consider I install litlellm through pip but I had the feling that poetry is used now; I tried in one terminal:
`poetry run litellm --model huggingface/bigcode/starcoder -c litellm_server/router_config.yaml --debug
`
and in a different one:
`
poetry run litellm --model gpt-3.5-turbo --test
`
but I am getting errors on both screens; the error message recommends to change the url to connect to https://my_ollama_api; the debug screen is large as some exceptions are fired from models.py:

> raise MissingSchema(requests.exceptions.MissingSchema: Invalid URL 'my_ollama_api_base/api/generate': No scheme supplied. Perhaps you meant https://my_ollama_api_base/api/generate?

**Reply by ghost** [reply]
I'm in PST - but happy to do whenever works best for you @shuther 

Also happy to set up a 1:1 support channel. 

Here's my details: 
LinkedIn: https://www.linkedin.com/in/krish-d/
Discord: Krrish#8748

**Reply by ghost** [reply]
we use pip

Here's the docs: https://docs.litellm.ai/docs/simple_proxy

Let me know if this works: 
`pip install litellm`
`export OPENAI_API_KEY = sk..`
`litellm --model gpt-3.5-turbo --test`

**Reply by shuther** [reply]
so one issue was that poetry did not load the .env by default, so `poetry self add poetry-dotenv-plugin` was required.
for the 2nd issue I will share it through discord

**Comment by ishaan-jaff** [comment]
we should deprecate our gpt cache support: https://github.com/BerriAI/litellm/blob/main/litellm/gpt_cache.py 

- I have not heard anyone using it 
- we now have our own caching support with redis

**Comment by ghost** [comment]
OpenAI v1 drops support for invalidrequesterror and now calls it badrequesterror. 
openai.InvalidRequestError -> openai.BadRequestError

**Comment by ghost** [comment]
OpenAI v1 changes how exception works 

- All exceptions inherit from API Error (same as before) 
- APIError now requires the request object (an httpx client) to be passed in

**Comment by kinthaiofficial** [comment]
A unified interface for LLM providers is exactly what multi-agent systems need at the model tier — agents shouldn't have to handle provider-specific APIs when their job is orchestration.

A few things worth getting right in v1.0 for agent use cases:

**Cost pass-through** — when LiteLLM wraps a provider call, the response should include the actual cost (not just token counts) in a standardized field. Agent systems need to do budget accounting, and having to maintain a separate cost lookup table for every provider is error-prone.

**Request attribution** — when the same LiteLLM instance is shared across agents, the call needs to carry `agent_id` and `delegation_chain_id` metadata that flows through to logging. Otherwise you can't distinguish which agent made which call at the model tier.

**Budget ceiling as a first-class parameter** — not just rate limits (requests/min) but spend ceilings ($X total for this call). If the estimated cost of the call exceeds the ceiling, refuse before making the API call rather than failing after. This is how you build economically safe agents.

**Streaming with cost events** — during streaming, emit partial cost events so the calling agent can see "I've spent $0.001 so far" rather than getting the full cost only at stream end. Useful for early termination if approaching budget.

The economic model that makes these features meaningful in a multi-agent context: https://blog.kinthai.ai/agent-wallet-economic-models-autonomous-agents

What's the top thing users are asking for in v1.0 beyond provider coverage?
